Adding 'dump timeseries' command

spex66 opened the pull request:
# Description

- added `DumpTimeSeriesCommand` through copy/paste search/replace approach from `DumpAssetCommand`
  - removed the `group` hierarchy-logic which caused some adoption (cannot think of a grouping feature, beside datasets?)
  - removed all `label` related code, as timeseries is not supporting labels
  - improved interactive mode 
    - listing datasets or assets with `clear-name (external_id) [count]`
    - added an "Abort" option

- TODO: (left TODO comments in code too)
  1. [x] naming convention timeseries, time_series, how to use singular vs plural [DONE]
  2. [x] assetId needs externalId replace [DONE]
    - so only storing `externalId` in `self.asset_by_id` not the whole `as_write()` object
  4. [x] ~~put it under ASSET feature flag support, let me know how to do it, maybe changing ASSETS feature to DUMP?~~
    - moved under TIMESERIES feature flag

After code-review looking into documentation, CHANGELOG, version, ..

## Checklist

- [ ] Tests added/updated.
- [x] Run Demo Job Locally.
  - [x] `dump timeseries --clean -f parquet --data-set WITSML` > 611k ts > resulted in two parquet files. Checked data in `visidata`.
  - [x]  `dump timeseries --clean -f csv --data-set WITSML --limit 1000` > checked in visidata
  - [x]  `dump timeseries --clean -f yaml --data-set WITSML --limit 1000` > checked in editor, but not loaded in total

```bash
➟  cdf-tk --env-path .env_prod_read dump timeseries --clean -f parquet --data-set WITSML
Retrieving time_series         ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00
Writing time_series to file(s) ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00
Dumped 611,761 time_series to tmp
Dumped 1 data sets to tmp/data_sets/time_series.DataSet.yaml
```

- [ ] Documentation updated. > not yet, is this required for FEATURES?
- [ ] Changelogs updated in [CHANGELOG.cdf-tk.md](https://github.com/cognitedata/toolkit/blob/main/CHANGELOG.cdf-tk.md).
- [ ] Template changelogs updated in [CHANGELOG.templates.md](https://github.com/cognitedata/toolkit/blob/main/CHANGELOG.templates.md).
- [ ] Version bumped. > not sure which one
  [_version.py](https://github.com/cognitedata/toolkit/blob/main/cognite/cognite_toolkit/_version.py) and
  [pyproject.toml](https://github.com/cognitedata/toolkit/blob/main/pyproject.toml) per [semantic versioning](https://semver.org/).

doctrino commented on cognite_toolkit/_cdf.py:
```diff
@@ -91,7 +91,9 @@
     )
 
 default_typer_kws = dict(
-    pretty_exceptions_short=False, pretty_exceptions_show_locals=False, pretty_exceptions_enable=False
+    pretty_exceptions_short=False,
+    pretty_exceptions_show_locals=False,
+    pretty_exceptions_enable=False,
```
It seems you have some extra automatic black linting on in your IDE that has a config we don't use in Toolkit. 

This clutters your PR and make it harder to read, so I would ideally like you to turn it off.

spex66 replied:
understandable, but that's tricky for me atm :/
- using black-formatter on save, which I thought is best-practice 
  - but which tends to reformat oneliners as I can see
- applying `pre-commit run` does not fix the format as wanted by the repo

I know you use pycharm, but which local formatter do you use?

spex66 replied:
deactivated my local `black` formatter in vscode until we find a solution

doctrino replied:
I don't use autoformatter in my IDE. Instead, I just allow the local pre-commit linting in each project I work determine the formatting. 

I have earlier experimented with having an autoformatter on save, and stopped for this exact reason.

doctrino commented on cognite_toolkit/_cdf.py:
```diff
@@ -1128,6 +1166,82 @@ def dump_asset_cmd(
             )
         )
 
+    @dump_app.command("timeseries")
```
I think this deserves its own feature flag, and not be with the same one as assets. Can you add another and add a if-block here?

spex66 replied:
Agree with separate FEATURE.

Missed before and found now `setup_asset_loader.py` and it is above my head to implement it for timeseries?! Which `_modify` steps to implement?

doctrino replied:
As suggested in the main comment, I think you should skip the modification of the loader. And instead just focus on the new dump functionality. 

This means that the feature this PR should provide is ONLY dump of timeseries, but not ability to write timeseries from parquet and csv to CDF. Lets instead consider that in 0.3.0a1.

doctrino commented on cognite_toolkit/_cdf_tk/prototypes/resource_loaders.py:
```diff
@@ -17,18 +17,197 @@
 
 
+@final
+class TimeSeriesLoader(ResourceLoader[str, TimeSeriesWrite, TimeSeries, TimeSeriesWriteList, TimeSeriesList]):
```
What is the motivation for implementing this one? We already have a TimeSeriesLoader in the core code base? 

Is it only to create a new load method? And allow parquet files and csv files? Then, we should dynamically only modify these methods on runtime when the flag is set. You can see how the file and timeseries loaders are modified in the `setup_asset_loader`. 

However, I am leaning towards waiting with this functionality and instead only implement the dump command and not modify the loader.

spex66 replied:
> I suggest, removing the `TimeSeriesLoader` you added and save support for loading parquet and csv until `0.3.0a1` (as it is a bit messy to implement runtime modification of classes as we do for prototypes).

> What is the motivation for implementing this one? 

It was a copy/paste approach mimic `dump_assets.py` / `AssetLoader` as template
```
from cognite_toolkit._cdf_tk.prototypes.resource_loaders import AssetLoader
```

Only the `folder_name` attribute is used, but I didn't want to mess up the symmetry (w/o knowing about all details).
What I can do instead is simply define `time_series_folder_name` locally in `dump_timeseries.py` and remove the `TimeSeriesLoader`

that's fine?

doctrino commented on cognite_toolkit/_cdf_tk/prototypes/commands/dump_timeseries.py:
```diff
@@ -0,0 +1,306 @@
+
+    def _get_asset_external_id(self, client: CogniteClient, asset_id: int) -> str:
+        if asset_id in self.asset_by_id:
+            return cast(str, self.asset_by_id[asset_id])
```
Is this cast necessary? It seems the `asset_by_id` is typed correctly.
```suggestion
            return self.asset_by_id[asset_id]
```

spex66 replied:
copy/pasted the `cast(str, self.data_set_by_id[data_set_id].external_id)` from dump_assets.py
- and I see now that `self.asset_by_id: dict[int, str] = {}` is explicit about the `str` type
- but `self.data_set_by_id: dict[int, DataSetWrite] = {}` not

doctrino commented on cognite_toolkit/_cdf_tk/prototypes/commands/dump_timeseries.py:
```diff
@@ -0,0 +1,306 @@
+                write_time_series.append(write)
+            yield write_time_series
+
+    def _get_time_series_external_id(self, client: CogniteClient, root_id: int) -> str:
```
I would call the parameter just `time_series_id` 
```suggestion
    def _get_time_series_external_id(self, client: CogniteClient, time_series_id: int) -> str:
```

doctrino replied:
Based on the comment above, not sure if this function is needed at all.

doctrino commented on cognite_toolkit/_cdf_tk/prototypes/commands/dump_timeseries.py:
```diff
@@ -0,0 +1,306 @@
+            # TODO: how to separate `times_series` as list from `time_series` as one item?
+            for one_time_series in time_series_list:
+                write = one_time_series.as_write().dump(camel_case=True)
+                write.pop("parentId", None)
```
No parents for time series
```suggestion
```

doctrino commented on cognite_toolkit/_cdf_tk/prototypes/commands/dump_timeseries.py:
```diff
@@ -0,0 +1,306 @@
+                    metadata = write.pop("metadata")
+                    for key, value in metadata.items():
+                        write[f"metadata.{key}"] = value
+                if "rootId" in write:
+                    root_id = write.pop("rootId")
+                    write["rootExternalId"] = self._get_time_series_external_id(client, root_id)
```
TimeSeries does not have a `rootId` as far as I can see
```suggestion
```

spex66 replied:
good catch
As a consequence I have to think about support for dedicated ts-extId (interactive, parameter or per file) and adopting the unused

- `def _get_time_series_external_id(self, client: CogniteClient, root_id: int) -> str:` 
for it 🤔

doctrino commented on cognite_toolkit/_cdf_tk/prototypes/commands/dump_timeseries.py:
```diff
@@ -0,0 +1,306 @@
+
+    def _buffer(self, time_series_iterator: Iterator[list[dict[str, Any]]]) -> Iterator[pd.DataFrame]:
+        """Iterates over time_series until the buffer reaches the filesize."""
+        stored_time_series: pd.DataFrame = pd.DataFrame()
```
Is `mypy` difficult on this one? Should not be necessary
```suggestion
        stored_time_series = pd.DataFrame()
```

doctrino commented on cognite_toolkit/_cdf_tk/prototypes/commands/dump_timeseries.py:
```diff
@@ -0,0 +1,306 @@
+from cognite_toolkit._cdf_tk.utils import CDFToolConfig
+
+#
+# TODO: what is the naming convention for TimeSeries, timeseries, time_series, singular or plural?
```
In the Python-SDK we use TimeSeries, and time_series. Unfortunately, there is no nice way to distinguish singular and plural, but can use Hungarian notation (as you have done) with `time_series_list` and `time_series_lists`.

spex66 replied:
thx for pointing `time_series_lists` out, helped to replace my odd `one_timee_series` name :)

doctrino reviewed (commented):
Overall good job. 

I suggest, removing the `TimeSeriesLoader` you added and save support for loading parquet and csv until `0.3.0a1` (as it is a bit messy to implement runtime modification of classes as we do for prototypes).

One more caveat here. This dump functionality (same for assets) is a prototype, thus there is no guarantee that it will be added to Toolkit. I think we should have a discussion after summer if this is a good for the scope of Toolkit, or if it will confuse from the core functionality.

spex66 commented:
next round:
- timeseries has own feature-flag now
- removed `TimeSeriesLoader`
- fixed (most) suggestions with deactivated local `black` formatter
- tested and still working :)

TODOs:
- how to implement a `setup_timeseries_loader.py`, which `_modify` steps and why?
- FYI: added a parameter `--timeseries / -t` to provide an external-id. Not implemented yet.
  - need a file-loader option too

spex66 commented:
pushed fixes and improved interactive mode (updated PR description)

doctrino commented on cognite_toolkit/_cdf_tk/commands/featureflag.py:
```diff
@@ -20,6 +20,10 @@ class Flags(Enum):
     INTERNAL: ClassVar[dict[str, Any]] = {"visible": False, "description": "Does nothing"}
     IMPORT_CMD: ClassVar[dict[str, Any]] = {"visible": True, "description": "Enables the import sub application"}
     ASSETS: ClassVar[dict[str, Any]] = {"visible": True, "description": "Enables the support for loading assets"}
+    TIMESERIES: ClassVar[dict[str, Any]] = {
+        "visible": True,
+        "description": "Enables the support for loading timeseries",
+    }
```
We already have support for loading timeseries. 
```suggestion
    TIMESERIES_DUMP: ClassVar[dict[str, Any]] = {
        "visible": True,
        "description": "Enables the support for dumping timeseries",
    }
```

doctrino commented:
> * how to implement a `setup_timeseries_loader.py`, which `_modify` steps and why?

No need to do any changes to the TimeSeries loader as you are only introducing a dump command.

doctrino commented on cognite_toolkit/_cdf_tk/prototypes/commands/dump_assets.py:
```diff
@@ -209,20 +263,26 @@ def _select_hierarchy_and_data_set(
 
             if what == "Done":
                 break
+            elif what == "Abort":
+                return [], []
```
Nice

doctrino commented on cognite_toolkit/_cdf_tk/prototypes/commands/dump_assets.py:
```diff
@@ -188,6 +191,57 @@ def _buffer(self, asset_iterator: Iterator[tuple[str, list[dict[str, Any]]]]) ->
+        """
+        assert isinstance(choice, questionary.Choice)
+        assert isinstance(choice.title, str)  # minimum external_id is set
+        return choice.title.lower()
```
Maybe worth a casefold in case of international characters? 

```suggestion
        return choice.title.casefold()
```

Example:
![image](https://github.com/user-attachments/assets/08920873-9c5a-4c39-9eb4-5b2059f297bf)

spex66 replied:
nice, didn't even know it exists!

spex66 replied:
would be interesting how it works with Japan projects 🤔 

asked copilot:
> Japanese characters do not have uppercase or lowercase forms.
> Therefore, applying [str.casefold()](vscode-file://vscode-app/c:/Program%20Files/Microsoft%20VS%20Code/resources/app/out/vs/code/electron-sandbox/workbench/workbench.html) to a Japanese string will not change the string. It will return the string as is.
```
japanese_string = "こんにちは"
casefolded_string = japanese_string.casefold()

print(casefolded_string)  # Output: こんにちは
```

doctrino commented on cognite_toolkit/_cdf_tk/prototypes/commands/dump_assets.py:
```diff
@@ -188,6 +191,57 @@ def _buffer(self, asset_iterator: Iterator[tuple[str, list[dict[str, Any]]]]) ->
+        `item_id` and `item` came in separate as item is DataSetWrite w/o `id`.
+        """
+
+        ts_count = None
```
Not used
```suggestion
```

doctrino commented on cognite_toolkit/_cdf_tk/prototypes/commands/dump_assets.py:
```diff
@@ -188,6 +191,57 @@ def _buffer(self, asset_iterator: Iterator[tuple[str, list[dict[str, Any]]]]) ->
+        """Using LRU decorator w/o limit instead of another lookup map."""
+        return client.assets.aggregate_count(advanced_filter=Equals("rootId", item_id))
+
+    @lru_cache
+    def get_choice_count(self, item_id: int, item: Asset | DataSetWrite, client: CogniteClient) -> int:
+        """Using LRU decorator w/o limit instead of another lookup map."""
+        if isinstance(item, DataSetWrite):
+            return client.time_series.aggregate_count(advanced_filter=Equals("dataSetId", item_id))
+        elif isinstance(item, Asset):
+            return client.assets.aggregate_count(advanced_filter=Equals("dataSetId", item_id))
+        else:
+            raise TypeError(f"Unsupported item type: {type(item)}")
```
This method is not used
```suggestion
```

doctrino commented on cognite_toolkit/_cdf_tk/prototypes/commands/dump_timeseries.py:
```diff
@@ -0,0 +1,342 @@
+from cognite_toolkit._cdf_tk.loaders import DataSetsLoader
+from cognite_toolkit._cdf_tk.utils import CDFToolConfig
+
+# from cognite_toolkit._cdf_tk.prototypes.resource_loaders import TimeSeriesLoader
+TIME_SERIES_FOLDER_NAME = "timeseries"
```
Suggest just getting the folder name from the TimeSeriesLoader in the core of the code base.
```suggestion
from cognite_toolkit._cdf_tk.loaders import TimeSeriesLoader
TIME_SERIES_FOLDER_NAME = TimeSeriesLoader.folder_name
```

doctrino reviewed (commented):
Great improvements. Very nice for asset dump as well with name external id and count. Very useful.

I think this PR is now close to be ready to merge, just a few more minor adjustments left.

doctrino commented on cognite_toolkit/_cdf_tk/prototypes/commands/dump_assets.py:
```diff
@@ -240,7 +229,7 @@ def _get_choice_title(self, choice: str | questionary.Choice | dict[str, Any]) -
         assert isinstance(choice, questionary.Choice)
         assert isinstance(choice.title, str)  # minimum external_id is set
-        return choice.title.lower()
+        return choice.title.casefold()  # superior to lower case like `ß>ss` in German
```
🤣
